### PR TITLE
bnorm+relu fuse for mkldnn (inference)

### PR DESCRIPTION
**Pull request by pzelazko-intel**
I've added batch norm + relu fuse case to inference_transpiler.
In next step, I'm going to create training transpiler doing same operation.

### CODE REVIEW

File: python/paddle/fluid/transpiler/inference_transpiler.py [hunk 1]
```diff
@@ -21,13 +22,13 @@
 class InferenceTranspiler:
     def transpile(self, program, place, scope=None):
         '''
-        Transpile the program. Support only fuse batch normalization now.
+        Transpile the program. Support only batch normalization and relu fuse now.
```
Comment (tensor-tang) [inline python/paddle/fluid/transpiler/inference_transpiler.py]: You mean batch norm and relu fuse in MKLDNN,

But the plain batch norm is fused here.

File: python/paddle/fluid/transpiler/inference_transpiler.py [hunk 2]
```diff
@@ -39,59 +40,98 @@ def transpile(self, program, place, scope=None):
...
+            if current_op.type in ['batch_norm']:
+                next_op = self.block.ops[i + 1]
+                if next_op.type == 'relu':
+                    # modify bnorm OP to include relu
+                    current_op.set_attr("fuse_with_relu", True)
+                    # remove relu OP
+                    self.block.remove_op(i + 1)
+            i = i + 1
+
+        # TODO(luotao): use clone() method to flush the program.desc in force,
```
Comment (tensor-tang) [inline python/paddle/fluid/transpiler/inference_transpiler.py]: @luotao1 please help review below.

File: python/paddle/fluid/layers/nn.py [hunk 3]
```diff
@@ -4799,9 +4802,7 @@ def image_resize_short(input, out_short_len, resample='BILINEAR'):
 
 def gather(input, index):
     """
-    **Gather Layer**
-
-    Output is obtained by gathering entries of the outer-most dimension 
+    Output is obtained by gathering entries of the outer-most dimension
```
Comment (luotao1) [inline python/paddle/fluid/layers/nn.py]: Seem that there is some diff in the annotation of `nn.py`, do you modify this file? If not, you could remain it.

File: benchmark/fluid/fluid_benchmark.py [hunk 4]
```diff
@@ -131,6 +131,10 @@ def train(avg_loss, infer_prog, optimizer, train_reader, test_reader, batch_acc,
     exe = fluid.Executor(place)
     exe.run(startup_prog)
 
+    # Use inference_transpiler to speedup
+    t = fluid.InferenceTranspiler()
+    t.transpile(infer_prog, place)
```
Comment (luotao1) [inline benchmark/fluid/fluid_benchmark.py]: It's not suitable to add transpiler function here, since its a `train` benchmark. @typhoonzero Do you have some better suggestion?

File: python/paddle/fluid/transpiler/inference_transpiler.py [hunk 5]
```diff
@@ -39,59 +40,98 @@ def transpile(self, program, place, scope=None):
             scope = global_scope()
         if not isinstance(scope, core.Scope):
             raise TypeError("scope should be as Scope type or None")
-        self.fuse_batch_norm(program, place, scope)
+        use_mkldnn = bool(os.getenv("FLAGS_use_mkldnn", False))
+        if not use_mkldnn:
+            self.fuse_batch_norm(program, place, scope)
+        else:
+            self.fuse_relu(program)
```
Comment (luotao1) [inline python/paddle/fluid/transpiler/inference_transpiler.py]: - fuse_batch_norm is also suitable for MKLDNN. 
- After `fuse_batch_norm`, `fuse_relu` is conv+relu, do you mean that?

File: python/paddle/fluid/layers/nn.py (hunk 3 above)
Comment (pzelazko-intel) [reply python/paddle/fluid/layers/nn.py]: Seems like I have unneceessary deleted "**Gather Layer**" - I will restore it.

File: python/paddle/fluid/transpiler/inference_transpiler.py (hunk 5 above)
Comment (pzelazko-intel) [reply python/paddle/fluid/transpiler/inference_transpiler.py]: 1) If `fuse_batch_norm` is suitable for MKLDNN, then I'll leave it without checking `use_mkldnn` flag.
2) fuse_relu deletes relu from "batch norm + relu" pair.

File: benchmark/fluid/fluid_benchmark.py (hunk 4 above)
Comment (pzelazko-intel) [reply benchmark/fluid/fluid_benchmark.py]: I could move it higher to `main`. However, `InferenceTranspiler` requires place as a parameter and extracting it happens in `train`. So the drawback would be that I would have to extract place again. Or I could add 'place' parameter to `train` and extract it in `main.

File: python/paddle/fluid/transpiler/inference_transpiler.py (hunk 5 above)
Comment (luotao1) [reply python/paddle/fluid/transpiler/inference_transpiler.py]: Since we have `fuse_batch_norm` already, there is no `batch_norm` op in inference program. Thus, what's the usage of `fuse_relu`?

File: paddle/fluid/operators/batch_norm_mkldnn_op.cc [hunk 6]
```diff
@@ -80,6 +80,7 @@ class BatchNormMKLDNNOpKernel : public paddle::framework::OpKernel<T> {
     const float epsilon = ctx.Attr<float>("epsilon");
     const float momentum = ctx.Attr<float>("momentum");
     const bool is_test = ctx.Attr<bool>("is_test");
+    const bool fuse_with_relu = ctx.Attr<bool>("fuse_with_relu");
```
Comment (luotao1) [inline paddle/fluid/operators/batch_norm_mkldnn_op.cc:69]: why should we add `fuse_with_relu` attribute here? 
```
t = fluid.InferenceTranspiler()
t.transpile(infer_prog, place)
```
is enough. Do you mean there is `mkldnn::fuse_bn_relu` function in MKLDNN?

Comment (pzelazko-intel) [reply paddle/fluid/operators/batch_norm_mkldnn_op.cc:69]: `mkldnn::fuse_bn_relu` is a flag for MKLDNN batch norm telling it to execute relu along with batch norm.

Comment (pzelazko-intel) [reply paddle/fluid/operators/batch_norm_mkldnn_op.cc:69]: If we execute `fuse_batch_norm` always (no `not ise_mkldnn` if), then `fuse_relu` makes sense only in case where there is no conv before batch norm. I don't know if such case in fact ever exists.

If no, then I can skip this PR and create similiar one for the training, which I have already completed.

Comment (luotao1) [reply paddle/fluid/operators/batch_norm_mkldnn_op.cc:69]: > where there is no conv before batch norm

For DenseNet https://github.com/liuzhuang13/DenseNet, there is BN+Relu+Conv, thus, `fuse_relu` is useful in this case.

> I can skip this PR and create similiar one for the training

`fuse_relu` for training is needed as well.

Comment (pzelazko-intel) [reply paddle/fluid/operators/batch_norm_mkldnn_op.cc:69]: So I assume this PR is OK. After it's merged, I'll create PR for training transpiler.

File: benchmark/fluid/fluid_benchmark.py (hunk 4 above)
Comment (luotao1) [reply benchmark/fluid/fluid_benchmark.py]: The benchmark of inference is on the original model now. Thus, how about remove line 134-136, or add an option to control it?

File: python/paddle/fluid/transpiler/inference_transpiler.py [hunk 7]
```diff
@@ -54,19 +55,57 @@ def transpile(self, program, place, scope=None):
...
+        while i < len(self.block.ops) - 1:
+            current_op = self.block.ops[i]
+            if current_op.type in ['batch_norm']:
+                next_op = self.block.ops[i + 1]
+                if next_op.type == 'relu':
+                    # modify bnorm OP to include relu
+                    current_op.set_attr("fuse_with_relu", True)
+                    # remove relu OP
+                    self.block.remove_op(i + 1)
+            i = i + 1
```
Comment (luotao1) [inline python/paddle/fluid/transpiler/inference_transpiler.py:99]: - could you give some unit test to validate the accuracy of `fuse_with_relu`?
- could you call `self._remove_unused_var `, since `remove_op` will not remove variables

File: python/paddle/fluid/transpiler/inference_transpiler.py [hunk 8]
```diff
@@ -20,12 +21,12 @@
 
 class InferenceTranspiler:
     '''
-    Convert the fluid program to optimized inference program. 
-    
-    There are several optimizations, only fuse batch normalization is supported now.
+    Convert the fluid program to optimized inference program.
+
+    There are several optimizations.
```
Comment (luotao1) [inline python/paddle/fluid/transpiler/inference_transpiler.py]: There are several optimizations:

- fuse convolution and batch normalization
- fuse batch normalization and relu (MKLDNN only)

File: python/paddle/fluid/transpiler/inference_transpiler.py [hunk 9]
```diff
@@ -54,19 +55,57 @@ def transpile(self, program, place, scope=None):
...
+        use_mkldnn = bool(os.getenv("FLAGS_use_mkldnn", False))
+        if use_mkldnn:
+            self.fuse_relu(program)
+
+    def fuse_relu(self, program):
+        '''
+        Transpile the program by fused relu activation for MKLDNN program.
+
+        Relu activation following batch norm OP can be fused by adding
+        'fuse_with_relu' attribute to batch norm OP.
```
Comment (luotao1) [inline python/paddle/fluid/transpiler/inference_transpiler.py]: 'fuse_with_relu' -> :math:`fuse_with_relu`

File: python/paddle/fluid/transpiler/inference_transpiler.py [hunk 10]
```diff
@@ -54,19 +55,57 @@ def transpile(self, program, place, scope=None):
...
+
+        The result of fuse is:
+            - before:
+                - batch_norm->relu->any_other_op
+            - after:
+                - batch_norm->any_other_op
```
Comment (luotao1) [inline python/paddle/fluid/transpiler/inference_transpiler.py]: The format of line 69-73 is not correct. You can use https://github.com/PaddlePaddle/FluidDoc to see the generated html. And paste the generated picture here like https://github.com/PaddlePaddle/Paddle/pull/11521.

If you have any question about how to generate the API reference, please feel open to ask me.

File: python/paddle/fluid/transpiler/inference_transpiler.py [hunk 11]
```diff
@@ -54,19 +55,57 @@ def transpile(self, program, place, scope=None):
         if not isinstance(scope, core.Scope):
             raise TypeError("scope should be as Scope type or None")
         self.fuse_batch_norm(program, place, scope)
+        use_mkldnn = bool(os.getenv("FLAGS_use_mkldnn", False))
+        if use_mkldnn:
+            self.fuse_relu(program)
+
+    def fuse_relu(self, program):
```
Comment (luotao1) [inline python/paddle/fluid/transpiler/inference_transpiler.py]: - how about the function name of `fuse_relu_mkldnn`?
- you may add the check of `FLAGS_use_mkldnn=True` in this function, thus, other people will not use it in plain CPU.

File: python/paddle/fluid/transpiler/inference_transpiler.py (hunk 7 above)
Comment (pzelazko-intel) [reply python/paddle/fluid/transpiler/inference_transpiler.py:99]: done

File: python/paddle/fluid/transpiler/inference_transpiler.py [hunk 12]
```diff
@@ -99,20 +148,20 @@ def fuse_batch_norm(self, program, place, scope):
...
         '''
         self.scope = scope
         self.place = place
         self.block = program.block(0)
-        self.input_map = {}  # store the input names should be adjusted 
+        self.input_map = {}  # store the input names should be adjusted
 
         i = 0
-        while i < len(self.block.ops):
+        while i < len(self.block.ops) - 2:
```
Comment (luotao1) [inline python/paddle/fluid/transpiler/inference_transpiler.py:159]: why do you change line 159?

Comment (pzelazko-intel) [reply python/paddle/fluid/transpiler/inference_transpiler.py:159]: Because in lines 174 and 183 we access `i+2` element.

**Review by luotao1 — APPROVED** [review]
LGTM
